libtsuba/src/ParquetReader.cpp

```diff
@@ -148,141 +120,251 @@ tsuba::ParquetReader::ReadFromUriSliced(const katana::Uri& uri) {
+      int64_t table_offset = row_offsets_[idx];
+      int64_t next_table_offset =
+          (idx == row_offsets_.size() - 1 ? std::numeric_limits<int64_t>::max()
+                                          : row_offsets_[idx + 1]);
+      std::shared_ptr<arrow::Table> table;
+      if (curr_global_row == table_offset &&
+          last_global_row >= next_table_offset) {
+        KATANA_CHECKED(EnsureReader(idx, true));
+        KATANA_CHECKED(readers_[idx]->ReadTable(&table));
+      } else {
+        KATANA_CHECKED(EnsureReader(idx, false));
+        table = KATANA_CHECKED(ReadTableSlice(
+            readers_[idx].get(), fvs_[idx].get(),
+            curr_global_row - table_offset,
+            std::min(
+                next_table_offset - table_offset,
+                last_global_row - table_offset)));
+      }
+      tables.emplace_back(std::move(table));
+      curr_global_row = next_table_offset;
+    }
+
+    return KATANA_CHECKED(arrow::ConcatenateTables(tables));
+  }
+
+  Result<std::shared_ptr<arrow::Table>> ReadTable(
+      std::vector<int32_t> col_indexes,
+      std::optional<tsuba::ParquetReader::Slice> slice = std::nullopt) {
+    if (slice) {
       return KATANA_ERROR(
-          ErrorCode::InvalidArgument,
-          "column index {} should be less than the number of columns {}", idx,
-          schema.num_fields());
+          katana::ErrorCode::NotImplemented,
+          "column subset read not implemented for slice! Sorry!");
     }
-    std::shared_ptr<arrow::ChunkedArray> column;
-    if (auto arr_it = read_arrays.find(idx); arr_it != read_arrays.end()) {
-      column = arr_it->second;
-    } else {
-      auto status = reader->ReadColumn(idx, &column);
-      if (!status.ok()) {
-        return KATANA_ERROR(
-            ErrorCode::ArrowError, "reading column {}: {}", idx, status);
+
+    std::vector<std::shared_ptr<arrow::Table>> tables;
+
+    for (auto& reader : readers_) {
+      std::vector<std::shared_ptr<arrow::Field>> fields;
+      std::vector<std::shared_ptr<arrow::ChunkedArray>> columns;
+      std::unordered_map<int32_t, std::shared_ptr<arrow::ChunkedArray>>
+          read_arrays;
+
+      std::shared_ptr<arrow::Schema> schema;
+      KATANA_CHECKED(reader->GetSchema(&schema));
+      for (int32_t idx : col_indexes) {
+        if (idx < 0) {
+          return KATANA_ERROR(
+              ErrorCode::InvalidArgument, "column indexes must be positive");
+        }
+        if (idx >= schema->num_fields()) {
+          return KATANA_ERROR(
+              ErrorCode::InvalidArgument,
+              "column index {} should be less than the number of columns {}",
+              idx, schema->num_fields());
+        }
+        std::shared_ptr<arrow::ChunkedArray> column;
+        if (auto arr_it = read_arrays.find(idx); arr_it != read_arrays.end()) {
+          column = arr_it->second;
+        } else {
+          auto status = reader->ReadColumn(idx, &column);
```

Comment:
I believe `KATANA_CHECKED` knows about arrow status as well, so you should be able to do: `KATANA_CHECKED(reader->ReadColumn(idx, &column));`